File: Assemblies/OS Consultant Tools/App Objects (Globals)/GBL_UI_Assembly/GBL Support Classes/GBL_SQL_Command_Builder.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Common;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using Microsoft.CSharp;
using Microsoft.Data.SqlClient;
using OneStream.Finance.Database;
using OneStream.Finance.Engine;
using OneStream.Shared.Common;
using OneStream.Shared.Database;
using OneStream.Shared.Engine;
using OneStream.Shared.Wcf;
using OneStream.Stage.Database;
using OneStream.Stage.Engine;
using OneStreamWorkspacesApi;
using OneStreamWorkspacesApi.V800;

namespace Workspace.__WsNamespacePrefix.__WsAssemblyName
{
    /// <summary>
    /// Generic SQL Command Builder that dynamically generates INSERT/UPDATE/DELETE commands
    /// based on DataTable schema. This eliminates hardcoded SQL statements and provides
    /// automatic synchronization when schema changes occur.
    /// </summary>
    public class GBL_SQL_Command_Builder
    {
        private readonly SqlConnection _connection;
        private readonly string _tableName;
        private readonly DataTable _dataTable;
        private readonly List<string> _primaryKeyColumns;
        private readonly List<string> _excludeFromUpdate;
        private readonly List<string> _excludeFromInsert;

        /// <summary>
        /// Initialize the SQL Command Builder
        /// </summary>
        /// <param name="connection">SQL connection to use</param>
        /// <param name="tableName">Name of the table to generate commands for</param>
        /// <param name="dataTable">DataTable with schema information</param>
        public GBL_SQL_Command_Builder(SqlConnection connection, string tableName, DataTable dataTable)
        {
            _connection = connection;
            _tableName = tableName;
            _dataTable = dataTable;
            _primaryKeyColumns = new List<string>();
            _excludeFromUpdate = new List<string>();
            _excludeFromInsert = new List<string>();
        }

        /// <summary>
        /// Set primary key column(s) for the table
        /// </summary>
        public void SetPrimaryKey(params string[] columnNames)
        {
            _primaryKeyColumns.Clear();
            _primaryKeyColumns.AddRange(columnNames);
        }

        /// <summary>
        /// Set columns to exclude from UPDATE operations (e.g., primary keys, create date/user)
        /// </summary>
        public void ExcludeFromUpdate(params string[] columnNames)
        {
            _excludeFromUpdate.Clear();
            _excludeFromUpdate.AddRange(columnNames);
        }

        /// <summary>
        /// Set columns to exclude from INSERT operations (e.g., identity columns)
        /// </summary>
        public void ExcludeFromInsert(params string[] columnNames)
        {
            _excludeFromInsert.Clear();
            _excludeFromInsert.AddRange(columnNames);
        }

        /// <summary>
        /// Build INSERT command based on DataTable schema
        /// </summary>
        public SqlCommand BuildInsertCommand(SqlTransaction transaction = null)
        {
            var columns = new List<string>();
            var parameters = new List<string>();

            foreach (DataColumn col in _dataTable.Columns)
            {
                if (_excludeFromInsert.Contains(col.ColumnName))
                    continue;

                columns.Add(col.ColumnName);
                parameters.Add($"@{col.ColumnName}");
            }

            var sql = $@"
                INSERT INTO {_tableName} (
                    {string.Join(", ", columns)}
                ) VALUES (
                    {string.Join(", ", parameters)}
                )";

            var command = transaction != null
                ? new SqlCommand(sql, _connection, transaction)
                : new SqlCommand(sql, _connection);

            command.UpdatedRowSource = UpdateRowSource.None;

            foreach (DataColumn col in _dataTable.Columns)
            {
                if (_excludeFromInsert.Contains(col.ColumnName))
                    continue;

                var param = command.Parameters.Add($"@{col.ColumnName}", GetSqlDbType(col.DataType));
                param.SourceColumn = col.ColumnName;

                if (col.MaxLength > 0 && (col.DataType == typeof(string)))
                {
                    param.Size = col.MaxLength;
                }
            }

            return command;
        }

        /// <summary>
        /// Build UPDATE command based on DataTable schema
        /// </summary>
        public SqlCommand BuildUpdateCommand(SqlTransaction transaction = null)
        {
            if (_primaryKeyColumns.Count == 0)
            {
                throw new InvalidOperationException("Primary key columns must be set before building UPDATE command");
            }

            var setClauses = new List<string>();
            var whereClauses = new List<string>();

            foreach (DataColumn col in _dataTable.Columns)
            {
                if (_primaryKeyColumns.Contains(col.ColumnName))
                {
                    whereClauses.Add($"{col.ColumnName} = @{col.ColumnName}");
                }
                else if (!_excludeFromUpdate.Contains(col.ColumnName))
                {
                    setClauses.Add($"{col.ColumnName} = @{col.ColumnName}");
                }
            }

            var sql = $@"
                UPDATE {_tableName} SET
                    {string.Join(",\n                    ", setClauses)}
                WHERE {string.Join(" AND ", whereClauses)}";

            var command = transaction != null
                ? new SqlCommand(sql, _connection, transaction)
                : new SqlCommand(sql, _connection);

            command.UpdatedRowSource = UpdateRowSource.None;

            // Add parameters for SET clause
            foreach (DataColumn col in _dataTable.Columns)
            {
                if (_primaryKeyColumns.Contains(col.ColumnName) || _excludeFromUpdate.Contains(col.ColumnName))
                    continue;

                var param = command.Parameters.Add($"@{col.ColumnName}", GetSqlDbType(col.DataType));
                param.SourceColumn = col.ColumnName;

                if (col.MaxLength > 0 && (col.DataType == typeof(string)))
                {
                    param.Size = col.MaxLength;
                }
            }

            // Add parameters for WHERE clause (primary keys)
            foreach (var pkColumn in _primaryKeyColumns)
            {
                var col = _dataTable.Columns[pkColumn];
                var param = command.Parameters.Add($"@{pkColumn}", GetSqlDbType(col.DataType));
                param.SourceColumn = pkColumn;
                param.SourceVersion = DataRowVersion.Original;

                if (col.MaxLength > 0 && (col.DataType == typeof(string)))
                {
                    param.Size = col.MaxLength;
                }
            }

            return command;
        }

        /// <summary>
        /// Build DELETE command based on primary key columns
        /// </summary>
        public SqlCommand BuildDeleteCommand(SqlTransaction transaction = null)
        {
            if (_primaryKeyColumns.Count == 0)
            {
                throw new InvalidOperationException("Primary key columns must be set before building DELETE command");
            }

            var whereClauses = new List<string>();

            foreach (var pkColumn in _primaryKeyColumns)
            {
                whereClauses.Add($"{pkColumn} = @{pkColumn}");
            }

            var sql = $@"
                DELETE FROM {_tableName}
                WHERE {string.Join(" AND ", whereClauses)}";

            var command = transaction != null
                ? new SqlCommand(sql, _connection, transaction)
                : new SqlCommand(sql, _connection);

            command.UpdatedRowSource = UpdateRowSource.None;

            foreach (var pkColumn in _primaryKeyColumns)
            {
                var col = _dataTable.Columns[pkColumn];
                var param = command.Parameters.Add($"@{pkColumn}", GetSqlDbType(col.DataType));
                param.SourceColumn = pkColumn;
                param.SourceVersion = DataRowVersion.Original;

                if (col.MaxLength > 0 && (col.DataType == typeof(string)))
                {
                    param.Size = col.MaxLength;
                }
            }

            return command;
        }

        /// <summary>
        /// Configure SqlDataAdapter with dynamically generated commands
        /// </summary>
        public void ConfigureAdapter(SqlDataAdapter adapter, SqlTransaction transaction = null)
        {
            adapter.InsertCommand = BuildInsertCommand(transaction);
            adapter.UpdateCommand = BuildUpdateCommand(transaction);
            adapter.DeleteCommand = BuildDeleteCommand(transaction);
            adapter.UpdateBatchSize = 0; // Set batch size for performance
        }

        /// <summary>
        /// Map .NET types to SQL Server types
        /// </summary>
        private SqlDbType GetSqlDbType(Type type)
        {
            if (type == typeof(int))
                return SqlDbType.Int;
            if (type == typeof(long))
                return SqlDbType.BigInt;
            if (type == typeof(short))
                return SqlDbType.SmallInt;
            if (type == typeof(byte))
                return SqlDbType.TinyInt;
            if (type == typeof(bool))
                return SqlDbType.Bit;
            if (type == typeof(DateTime))
                return SqlDbType.DateTime;
            if (type == typeof(decimal))
                return SqlDbType.Decimal;
            if (type == typeof(double))
                return SqlDbType.Float;
            if (type == typeof(float))
                return SqlDbType.Real;
            if (type == typeof(Guid))
                return SqlDbType.UniqueIdentifier;
            if (type == typeof(byte[]))
                return SqlDbType.VarBinary;
            if (type == typeof(string))
                return SqlDbType.NVarChar;

            return SqlDbType.NVarChar; // Default fallback
        }
    }
}

File: Assemblies/OS Consultant Tools/App Objects (Globals)/GBL_UI_Assembly/GBL Support Classes/GBL_SQL_Command_Builder.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Common
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports Microsoft.Data.SqlClient
Imports OneStream.Finance.Database
Imports OneStream.Finance.Engine
Imports OneStream.Shared.Common
Imports OneStream.Shared.Database
Imports OneStream.Shared.Engine
Imports OneStream.Shared.Wcf
Imports OneStream.Stage.Database
Imports OneStream.Stage.Engine
Imports OneStreamWorkspacesApi
Imports OneStreamWorkspacesApi.V800

Namespace Workspace.__WsNamespacePrefix.__WsAssemblyName
	''' <summary>
	''' Generic SQL Command Builder that dynamically generates INSERT/UPDATE/DELETE commands
	''' based on DataTable schema. This eliminates hardcoded SQL statements and provides
	''' automatic synchronization when schema changes occur.
	''' 
	''' This class works with any SqlConnection, including those created from:
	''' - BRApi.Database.CreateApplicationDbConnInfo(si)
	''' - BRApi.Database.CreateMergeDbConnInfo(si)
	''' - Any other database connection type
	''' </summary>
	Public Class GBL_SQL_Command_Builder
		Private ReadOnly _connection As SqlConnection
		Private ReadOnly _tableName As String
		Private ReadOnly _dataTable As DataTable
		Private ReadOnly _primaryKeyColumns As List(Of String)
		Private ReadOnly _excludeFromUpdate As List(Of String)
		Private ReadOnly _excludeFromInsert As List(Of String)

		''' <summary>
		''' Initialize the SQL Command Builder
		''' </summary>
		''' <param name="connection">SQL connection to use (can be from App DB or Merge DB)</param>
		''' <param name="tableName">Name of the table to generate commands for</param>
		''' <param name="dataTable">DataTable with schema information</param>
		Public Sub New(connection As SqlConnection, tableName As String, dataTable As DataTable)
			_connection = connection
			_tableName = tableName
			_dataTable = dataTable
			_primaryKeyColumns = New List(Of String)()
			_excludeFromUpdate = New List(Of String)()
			_excludeFromInsert = New List(Of String)()
		End Sub

		''' <summary>
		''' Set primary key column(s) for the table
		''' </summary>
		Public Sub SetPrimaryKey(ParamArray columnNames() As String)
			_primaryKeyColumns.Clear()
			_primaryKeyColumns.AddRange(columnNames)
		End Sub

		''' <summary>
		''' Set columns to exclude from UPDATE operations (e.g., primary keys, create date/user)
		''' </summary>
		Public Sub ExcludeFromUpdate(ParamArray columnNames() As String)
			_excludeFromUpdate.Clear()
			_excludeFromUpdate.AddRange(columnNames)
		End Sub

		''' <summary>
		''' Set columns to exclude from INSERT operations (e.g., identity columns)
		''' </summary>
		Public Sub ExcludeFromInsert(ParamArray columnNames() As String)
			_excludeFromInsert.Clear()
			_excludeFromInsert.AddRange(columnNames)
		End Sub

		''' <summary>
		''' Build INSERT command based on DataTable schema
		''' </summary>
		Public Function BuildInsertCommand(Optional transaction As SqlTransaction = Nothing) As SqlCommand
			Dim columns As New List(Of String)()
			Dim parameters As New List(Of String)()

			For Each col As DataColumn In _dataTable.Columns
				If _excludeFromInsert.Contains(col.ColumnName) Then
					Continue For
				End If

				columns.Add(col.ColumnName)
				parameters.Add("@" & col.ColumnName)
			Next

			Dim sql As String = _
				"INSERT INTO " & _tableName & " (" & Environment.NewLine & _
				"    " & String.Join(", ", columns) & Environment.NewLine & _
				") VALUES (" & Environment.NewLine & _
				"    " & String.Join(", ", parameters) & Environment.NewLine & _
				")"

			Dim command As SqlCommand
			If transaction IsNot Nothing Then
				command = New SqlCommand(sql, _connection, transaction)
			Else
				command = New SqlCommand(sql, _connection)
			End If

			command.UpdatedRowSource = UpdateRowSource.None

			For Each col As DataColumn In _dataTable.Columns
				If _excludeFromInsert.Contains(col.ColumnName) Then
					Continue For
				End If

				Dim param = command.Parameters.Add("@" & col.ColumnName, GetSqlDbType(col.DataType))
				param.SourceColumn = col.ColumnName

				If col.MaxLength > 0 AndAlso col.DataType Is GetType(String) Then
					param.Size = col.MaxLength
				End If
			Next

			Return command
		End Function

		''' <summary>
		''' Build UPDATE command based on DataTable schema
		''' </summary>
		Public Function BuildUpdateCommand(Optional transaction As SqlTransaction = Nothing) As SqlCommand
			If _primaryKeyColumns.Count = 0 Then
				Throw New InvalidOperationException("Primary key columns must be set before building UPDATE command")
			End If

			Dim setClauses As New List(Of String)()
			Dim whereClauses As New List(Of String)()

			For Each col As DataColumn In _dataTable.Columns
				If _primaryKeyColumns.Contains(col.ColumnName) Then
					whereClauses.Add(col.ColumnName & " = @" & col.ColumnName)
				ElseIf Not _excludeFromUpdate.Contains(col.ColumnName) Then
					setClauses.Add(col.ColumnName & " = @" & col.ColumnName)
				End If
			Next

			Dim sql As String = _
				"UPDATE " & _tableName & " SET" & Environment.NewLine & _
				"    " & String.Join("," & Environment.NewLine & "    ", setClauses) & Environment.NewLine & _
				"WHERE " & String.Join(" AND ", whereClauses)

			Dim command As SqlCommand
			If transaction IsNot Nothing Then
				command = New SqlCommand(sql, _connection, transaction)
			Else
				command = New SqlCommand(sql, _connection)
			End If

			command.UpdatedRowSource = UpdateRowSource.None

			' Add parameters for SET clause
			For Each col As DataColumn In _dataTable.Columns
				If _primaryKeyColumns.Contains(col.ColumnName) OrElse _excludeFromUpdate.Contains(col.ColumnName) Then
					Continue For
				End If

				Dim param = command.Parameters.Add("@" & col.ColumnName, GetSqlDbType(col.DataType))
				param.SourceColumn = col.ColumnName

				If col.MaxLength > 0 AndAlso col.DataType Is GetType(String) Then
					param.Size = col.MaxLength
				End If
			Next

			' Add parameters for WHERE clause (primary keys)
			For Each pkColumn As String In _primaryKeyColumns
				Dim col As DataColumn = _dataTable.Columns(pkColumn)
				Dim param = command.Parameters.Add("@" & pkColumn, GetSqlDbType(col.DataType))
				param.SourceColumn = pkColumn
				param.SourceVersion = DataRowVersion.Original

				If col.MaxLength > 0 AndAlso col.DataType Is GetType(String) Then
					param.Size = col.MaxLength
				End If
			Next

			Return command
		End Function

		''' <summary>
		''' Build DELETE command based on primary key columns
		''' </summary>
		Public Function BuildDeleteCommand(Optional transaction As SqlTransaction = Nothing) As SqlCommand
			If _primaryKeyColumns.Count = 0 Then
				Throw New InvalidOperationException("Primary key columns must be set before building DELETE command")
			End If

			Dim whereClauses As New List(Of String)()

			For Each pkColumn As String In _primaryKeyColumns
				whereClauses.Add(pkColumn & " = @" & pkColumn)
			Next

			Dim sql As String = _
				"DELETE FROM " & _tableName & Environment.NewLine & _
				"WHERE " & String.Join(" AND ", whereClauses)

			Dim command As SqlCommand
			If transaction IsNot Nothing Then
				command = New SqlCommand(sql, _connection, transaction)
			Else
				command = New SqlCommand(sql, _connection)
			End If

			command.UpdatedRowSource = UpdateRowSource.None

			For Each pkColumn As String In _primaryKeyColumns
				Dim col As DataColumn = _dataTable.Columns(pkColumn)
				Dim param = command.Parameters.Add("@" & pkColumn, GetSqlDbType(col.DataType))
				param.SourceColumn = pkColumn
				param.SourceVersion = DataRowVersion.Original

				If col.MaxLength > 0 AndAlso col.DataType Is GetType(String) Then
					param.Size = col.MaxLength
				End If
			Next

			Return command
		End Function

		''' <summary>
		''' Configure SqlDataAdapter with dynamically generated commands
		''' </summary>
		Public Sub ConfigureAdapter(adapter As SqlDataAdapter, Optional transaction As SqlTransaction = Nothing)
			adapter.InsertCommand = BuildInsertCommand(transaction)
			adapter.UpdateCommand = BuildUpdateCommand(transaction)
			adapter.DeleteCommand = BuildDeleteCommand(transaction)
			adapter.UpdateBatchSize = 0 ' Set batch size for performance
		End Sub

		''' <summary>
		''' Map .NET types to SQL Server types
		''' </summary>
		Private Function GetSqlDbType(dataType As Type) As SqlDbType
			If dataType Is GetType(Integer) Then
				Return SqlDbType.Int
			ElseIf dataType Is GetType(Long) Then
				Return SqlDbType.BigInt
			ElseIf dataType Is GetType(Short) Then
				Return SqlDbType.SmallInt
			ElseIf dataType Is GetType(Byte) Then
				Return SqlDbType.TinyInt
			ElseIf dataType Is GetType(Boolean) Then
				Return SqlDbType.Bit
			ElseIf dataType Is GetType(DateTime) Then
				Return SqlDbType.DateTime
			ElseIf dataType Is GetType(Decimal) Then
				Return SqlDbType.Decimal
			ElseIf dataType Is GetType(Double) Then
				Return SqlDbType.Float
			ElseIf dataType Is GetType(Single) Then
				Return SqlDbType.Real
			ElseIf dataType Is GetType(Guid) Then
				Return SqlDbType.UniqueIdentifier
			ElseIf dataType Is GetType(Byte()) Then
				Return SqlDbType.VarBinary
			ElseIf dataType Is GetType(String) Then
				Return SqlDbType.NVarChar
			Else
				Return SqlDbType.NVarChar ' Default fallback
			End If
		End Function
	End Class
End Namespace
